File: backend/src/main/java/com/example/backend/global/advice/GlobalControllerAdvice.java
package com.example.backend.global.advice;

import java.util.ArrayList;
import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import com.example.backend.domain.admin.exception.AdminException;
import com.example.backend.domain.category.exception.CategoryException;
import com.example.backend.domain.group.exception.GroupException;
import com.example.backend.domain.member.exception.MemberException;
import com.example.backend.domain.voter.exception.VoterException;
import com.example.backend.global.exception.GlobalErrorCode;
import com.example.backend.global.exception.GlobalException;
import com.example.backend.global.response.ErrorResponse;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.ConstraintViolationException;

/**
 * GlobalControllerAdvice
 * Global 예외 처리 advice 클래스
 * @author 100minha
 */
@ControllerAdvice
public class GlobalControllerAdvice {

	/**
	 * 컨트롤러 @Valid 검증 실패 예외 핸들러 메소드
	 * 컨트롤러 레이어에서
	 */
	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex,
		HttpServletRequest request) {
		List<ErrorResponse.ValidationError> errors = new ArrayList<>();

		ex.getBindingResult().getAllErrors().forEach(error -> {
			errors.add(ErrorResponse.ValidationError.builder()
				.field(((FieldError)error).getField())
				.message(error.getDefaultMessage())
				.build());
		});

		return ResponseEntity.status(ex.getStatusCode().value()).body(
			ErrorResponse.of(
				GlobalErrorCode.NOT_VALID.getMessage(),
				GlobalErrorCode.NOT_VALID.getCode(),
				request.getRequestURI(),
				errors
			)
		);
	}

	@ExceptionHandler(ConstraintViolationException.class)
	public ResponseEntity<ErrorResponse> handleConstraintViolationException(ConstraintViolationException ex,
		HttpServletRequest request) {
		List<ErrorResponse.ValidationError> errors = new ArrayList<>();

		ex.getConstraintViolations().forEach(violation -> {
			String fieldName = violation.getPropertyPath().toString();
			String message = violation.getMessage();
			errors.add(new ErrorResponse.ValidationError(fieldName, message));
		});

		return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
			ErrorResponse.of(
				GlobalErrorCode.NOT_VALID.getMessage(),
				GlobalErrorCode.NOT_VALID.getCode(),
				request.getRequestURI(),
				errors
			)
		);
	}

	@ExceptionHandler(GroupException.class)
	public ResponseEntity<ErrorResponse> handleGroupException(GroupException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(GlobalException.class)
	public ResponseEntity<ErrorResponse> handleGlobalException(GlobalException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(AdminException.class)
	public ResponseEntity<ErrorResponse> handleAdminException(AdminException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(CategoryException.class)
	public ResponseEntity<ErrorResponse> handleCategoryException(CategoryException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(MemberException.class)
	public ResponseEntity<ErrorResponse> handleMemberException(MemberException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(VoterException.class)
	public ResponseEntity<ErrorResponse> handleVoterException(VoterException ex, HttpServletRequest request) {
		return ResponseEntity.status(ex.getStatus())
			.body(ErrorResponse.of(ex.getMessage(), ex.getCode(), request.getRequestURI()));
	}

	@ExceptionHandler(Exception.class)
	public ResponseEntity<ErrorResponse> handleVoterException(Exception ex, HttpServletRequest request) {
		return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
			.body(ErrorResponse.of(ex.getMessage(), "500", request.getRequestURI()));
	}
}

File: backend/src/main/java/com/example/backend/global/advice/GlobalControllerAdvice.kt
package com.example.backend.global.advice

import com.example.backend.domain.admin.exception.AdminException
import com.example.backend.domain.category.exception.CategoryException
import com.example.backend.domain.group.exception.GroupException
import com.example.backend.domain.member.exception.MemberException
import com.example.backend.domain.voter.exception.VoterException
import com.example.backend.global.exception.GlobalErrorCode
import com.example.backend.global.exception.GlobalException
import com.example.backend.global.response.ErrorResponse
import com.example.backend.global.response.ErrorResponse.Companion.of
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolation
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import org.springframework.validation.ObjectError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.util.function.Consumer

/**
 * GlobalControllerAdvice
 * Global 예외 처리 advice 클래스
 * @author 100minha
 */
// TODO: !! 연산자 사용 제거
@ControllerAdvice
class GlobalControllerAdvice {
    /**
     * 컨트롤러 @Valid 검증 실패 예외 핸들러 메소드
     * 컨트롤러 레이어에서
     */
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(
        ex: MethodArgumentNotValidException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        val errors: MutableList<ErrorResponse.ValidationError> = ArrayList()

        ex.bindingResult.allErrors.forEach(Consumer { error: ObjectError ->
            errors.add(
                ErrorResponse.ValidationError(
                    (error as FieldError).field,
                    error.getDefaultMessage()
                )
            )
        })

        return ResponseEntity.status(ex.statusCode.value()).body(
            ErrorResponse.of(
                GlobalErrorCode.NOT_VALID.message,
                GlobalErrorCode.NOT_VALID.code,
                request.requestURI,
                errors
            )
        )
    }

    @ExceptionHandler(ConstraintViolationException::class)
    fun handleConstraintViolationException(
        ex: ConstraintViolationException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        val errors: MutableList<ErrorResponse.ValidationError> = ArrayList()

        ex.constraintViolations.forEach(Consumer { violation: ConstraintViolation<*> ->
            val fieldName = violation.propertyPath.toString()
            val message = violation.message
            errors.add(ErrorResponse.ValidationError(fieldName, message))
        })

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
            ErrorResponse.of(
                GlobalErrorCode.NOT_VALID.message,
                GlobalErrorCode.NOT_VALID.code,
                request.requestURI,
                errors
            )
        )
    }

    @ExceptionHandler(GroupException::class)
    fun handleGroupException(ex: GroupException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message!!, ex.code, request.requestURI))
    }

    @ExceptionHandler(GlobalException::class)
    fun handleGlobalException(ex: GlobalException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message, ex.code, request.requestURI))
    }

    @ExceptionHandler(AdminException::class)
    fun handleAdminException(ex: AdminException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message!!, ex.code, request.requestURI))
    }

    @ExceptionHandler(CategoryException::class)
    fun handleCategoryException(ex: CategoryException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message!!, ex.code, request.requestURI))
    }

    @ExceptionHandler(MemberException::class)
    fun handleMemberException(ex: MemberException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message, ex.code, request.requestURI))
    }

    @ExceptionHandler(VoterException::class)
    fun handleVoterException(ex: VoterException, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(ex.status)
            .body(ErrorResponse.of(ex.message!!, ex.code, request.requestURI))
    }

    @ExceptionHandler(Exception::class)
    fun handleVoterException(ex: Exception, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(ErrorResponse.of(ex.message ?: "INTERNAL_SERVER_ERROR", "500", request.requestURI))
    }
}
